### ISSUE TITLE
[Elasticsearch] has_child query support

### ISSUE THREAD

**Issue by dkarlovi** [rollerworks/search#245]
**Description**  
Elasticsearch allows for rudimentary parent/child relations with [`join`](https://www.elastic.co/guide/en/elasticsearch/reference/current/parent-join.html) data type. It allows specifying if some document is a child of some other document.

An example would be to find an `Article` document by both the content of its title and the content of its comments.

I propose to introduce support for new field type which would result in a [`has_child`](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-has-child-query.html) Elasticsearch query.

(This would, AFAIK, be the first Elasticsearch-specific type?)

**Example**  

Documents
```json
[
    {"id": 1, "type": "article", "title": "Eiffel Tower", "content": "Visit it."},
    {"id": 2, "type": {"name": "comment", "parent": 1}, "content": "What a lovely place!"}
]
```

Query
```json
{
    "query": {
        "has_child": {
            "type" : "comment",
            "query": {
                "match": {
                    "content": "place"
                }
            }
        }
    }
}
```

So the field would generate this part:

```
{
    "query": {
        "has_child": {
            "type" : $type,
            "query": $query
        }
    }
}
```

- `$type` - name of the child type
- `$query` - the usual query

Proposed mapping: `/$index/$type#$type>$query`, the `>` separates the `$type` and the `$query`, (taken from CSS selector, allowing us to support `has_parent` with `<`).

I'm mostly leaving this comment for myself since I plan to start working on this ASAP.

**Comment by dkarlovi** [comment]
@sstok do you have any objections here?

**Comment by dkarlovi** [comment]
Closed with #248
